executor/common_linux.h

```diff
@@ -1570,6 +1570,310 @@ static long syz_init_net_socket(volatile long domain, volatile long type, volati
+
+	struct hci_dev_stats stat;
+};
+
+struct hci_dev_req {
+	uint16 dev_id;
+	uint32 dev_opt;
+};
+
+#define HCISETSCAN _IOW('H', 208, int)
+#define HCIGETDEVINFO _IOR('H', 211, int)
+
+#define BUF_SIZE 4096
+#define ACL_HANDLE 256
+#define HCI_DEV 0
+
+static int vhci_fd = -1;
+static int vhci_init = 0;
+
+static long syz_init_net_socket(volatile long domain, volatile long type, volatile long proto);
+
+static int hci_send_event_packet(int fd, uint8 evt, void* data, size_t data_len)
+{
+	struct iovec iv[3];
+
+	struct hci_event_hdr hdr;
+	hdr.evt = evt;
+	hdr.plen = data_len;
+
+	uint8 type = HCI_EVENT_PKT;
+
+	iv[0].iov_base = &type;
+	iv[0].iov_len = sizeof(type);
+	iv[1].iov_base = &hdr;
+	iv[1].iov_len = sizeof(hdr);
+	iv[2].iov_base = data;
+	iv[2].iov_len = data_len;
+
+	return writev(fd, iv, sizeof(iv) / sizeof(struct iovec));
+}
+
+static int hci_send_event_cmd_complete(int fd, uint16 opcode, void* data, size_t data_len)
+{
+	struct iovec iv[4];
+
+	struct hci_event_hdr hdr;
+	hdr.evt = HCI_EV_CMD_COMPLETE;
+	hdr.plen = sizeof(struct hci_ev_cmd_complete) + data_len;
+
+	struct hci_ev_cmd_complete evt_hdr;
+	evt_hdr.ncmd = 1;
+	evt_hdr.opcode = opcode;
+
+	uint8 type = HCI_EVENT_PKT;
+
+	iv[0].iov_base = &type;
+	iv[0].iov_len = sizeof(type);
+	iv[1].iov_base = &hdr;
+	iv[1].iov_len = sizeof(hdr);
+	iv[2].iov_base = &evt_hdr;
+	iv[2].iov_len = sizeof(evt_hdr);
+	iv[3].iov_base = data;
+	iv[3].iov_len = data_len;
+
+	return writev(fd, iv, sizeof(iv) / sizeof(struct iovec));
+}
+
+static void process_command_pkt(int fd, char* buf, size_t buf_size)
+{
+	struct hci_command_hdr* hdr = (struct hci_command_hdr*)(buf + 1);
+
+	switch (hdr->opcode) {
+	case HCI_OP_READ_BD_ADDR: {
+		struct hci_rp_read_bd_addr rp;
+		rp.status = 0;
+		memcpy(&rp.bdaddr, "\x66\x55\x44\x33\x22\x11", 6);
+		hci_send_event_cmd_complete(fd, hdr->opcode, &rp, sizeof(rp));
+		return;
+	}
+	case HCI_OP_READ_BUFFER_SIZE: {
+		struct hci_rp_read_buffer_size rp;
+		rp.status = 0;
+		rp.acl_mtu = 1021;
+		rp.sco_mtu = 96;
+		rp.acl_max_pkt = 4;
+		rp.sco_max_pkt = 6;
+		hci_send_event_cmd_complete(fd, hdr->opcode, &rp, sizeof(rp));
+		return;
+	}
+	}
+
+	char status[0xf9] = {0};
+	hci_send_event_cmd_complete(fd, hdr->opcode, status, sizeof(status));
+}
+
+static void* event_thread(void* arg)
+{
+	int epoll_fd = *(int*)arg;
+
+	while (!vhci_init) {
+		struct epoll_event event;
+		epoll_pwait(epoll_fd, &event, 1, -1, NULL);
+
+		char buf[BUF_SIZE];
+		size_t buf_size = read(event.data.fd, buf, sizeof(buf));
+		if (buf[0] == HCI_COMMAND_PKT)
+			process_command_pkt(event.data.fd, buf, buf_size);
+	}
+
+	return NULL;
+}
+
+static void initialize_vhci()
+{
+#if SYZ_EXECUTOR
+	if (!flag_vhci_injection)
+		return;
+#endif
+
+	vhci_fd = open("/dev/vhci", O_RDWR);
+	if (vhci_fd == -1)
+		fail("open /dev/vhci failed");
+
+	int hci_socket = syz_init_net_socket(AF_BLUETOOTH, SOCK_RAW, 1);
+	if (hci_socket < 0)
+		fail("syz_init_net_socket failed");
+
+	int epoll_fd = epoll_create1(0);
```

Comment:
The epoll descriptor is used only in the thread, so it's better to create it there and close it there (currently it's not closed anywhere).

Reply:
I'm keeping this in this subroutine, so we can close it and as such, trigger pthread termination.